Cloud Production Checklist updates (#17953)
cockroachdb/docs: src/current/cockroachcloud/production-checklist.md

docs_area: deploy
---

Before deploying CockroachDB {{ site.data.products.cloud }} in production, it is important to understand the Shared Responsibility Model that delineates the responsibilities of Cockroach Labs and the customer in managing CockroachDB {{ site.data.products.cloud }} clusters. 

Under the Shared Responsibility Model, Cockroach Labs is responsible for the following tasks: 

- Cluster and cloud service availability and reliability
- Maintenance and security of hardware and operating systems

> - Database upgrades and security patches 

Reviewer: Maybe remove the word "upgrades" as this blurs the line between patching and upgrading

> - Daily and hourly cluster backups 

Reviewer 2: Might want to make this generic as the frequency of these backups may/will change over time

The customer is responsible for the following tasks:

> - Sizing the workload correctly

Reviewer 2: Is it sizing the workload or it sizing the cluster to fit the workload?

Reviewer: Estimating the workload volumetrics as accurately as possible?

- Scaling clusters based on workload
- Ensuring sufficient disk, compute, and memory capacity for each cluster

> - Monitoring cluster health and application performance

Reviewer 2: I don't feel strongly here, but are people able to delineate between "cluster health" and "hardware/os help" listed above? I feel like honestly this is fine as worded to start but more curious what you think here. App perf for sure is their responsibility. I am SO HAPPY we're explicitly stating this. SRE == us, DBA + App Dev == you

Reviewer: I believe there is room for misinterpretation here. Some customers will ask "don't you manage the cluster as it's a managed service?"

Reviewer: Maybe add a point below - something like:
Ensuring that the workload is distributed appropriately across the nodes of the cluster

Author: @alistair-roacher IIRC, we want the users to know that even if we manage the cluster, they are still responsible for monitoring the cluster health and app performance, right?

Reviewer: @Amruta-Ranade - Yes absolutely. We will monitor the cluster from an availability perspective but the customer should monitor all other aspects of cluster health.

- Performance tuning of SQL queries and schema
- Initiating major version upgrades and selecting maintenance windows for patch releases 
- Taking customer-owned backups (optional)

This page provides important recommendations for CockroachDB {{ site.data.products.cloud }} production tasks for which  the customer is responsible.

## Deployment options

> When planning your deployment, it is important to carefully review and choose the [deployment options](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/choose-a-deployment-option) that best meet your scale, cost, and security requirements. 

Reviewer: "best meet your scale, cost, and security requirements" - maybe mention resilience here?

Make sure your cluster has sufficient storage, CPU, and memory to handle the workload. The general formula to calculate the storage requirement is as follows:
 
`raw data (storage, in GB) * replication factor (3 by default) * remove 40% to account for compression (0.6) * headroom (1.5-2)`

For an example, see [Plan your Dedicated cluster]({% link cockroachcloud/plan-your-cluster.md %}).

## Topology patterns

When planning your deployment, it is important to carefully review and choose the [topology patterns](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/topology-patterns) that best meet your latency and resiliency requirements. This is especially crucial for multi-region deployments.

## Cluster management

You can create and manage CockroachDB {{ site.data.products.cloud }} clusters using the [Cloud Console](http://cockroachlabs.cloud), [Cloud API]({% link cockroachcloud/cloud-api.md %}), [ccloud CLI]({% link cockroachcloud/ccloud-get-started.md %}), or the [Terraform provider]({% link cockroachcloud/provision-a-cluster-with-terraform.md %}).

## Network authorization

> CockroachDB {{ site.data.products.cloud }} requires you to authorize the networks that can access the cluster in order to prevent denial-of-service and brute force password attacks. During the application development phase, you might have authorized only your local machine’s network. To move into production, you need to authorize your application server’s network.

Reviewer: "need to authorize your application server’s network" - this makes it sound like customers typically only have a single application server (which is not the case for our ICP)

## Use a pool of persistent connections

Creating the appropriate size pool of connections is critical to gaining maximum performance in an application. Too few connections in the pool will result in high latency as each operation waits for a connection to open up. But adding too many connections to the pool can also result in high latency as each connection thread is being run in parallel by the system. The time it takes for many threads to complete in parallel is typically higher than the time it takes a smaller number of threads to run sequentially.

For guidance on sizing, validating, and using connection pools with CockroachDB, see [Use Connection Pools](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/connection-pooling).

## Keeping connections current

After an application establishes a connection to CockroachDB {{ site.data.products.cloud }}, those connections can occasionally become invalid. This could be due to changes in the cluster topography, rolling [upgrades]({% link cockroachcloud/upgrade-policy.md %}) and restarts, network disruptions, or cloud infrastructure unavailability.

> Make sure connection validation and retry logic is used by your application. Validating and retrying connections is typically handled by the driver, framework, or the connection pool used by an application. For guidance on connection pool sizing, connection validation, and connection retry logic, see [Use Connection Pools](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/connection-pooling).

Reviewer 2: Do we have docs on connection retry logic specifically? We should definitely link that here -- it's a specific error code a lot of apps don't add to their retry logic.

Additionally, I know our connection pooling docs talk about setting max lifetime to 5min. Wonder if that's worth calling out explicitly here about why -- since a lot of customers don't and as a result their connections get severed by drain and then customers get sad when it adds latency / increases transaction errors

Author: Opened a tracking issue for error handling doc improvements: https://cockroachlabs.atlassian.net/browse/DOC-9237

> ## PCI ready features (Dedicated advanced)

Reviewer 2: How did you decide how to order the checklist? Tiny nit but it's confusing if everything below is Dedicated Advanced or just this. It's odd having it in the middle for that reason but there might be other reasons to keep it here. I don't feel strongly -- just one viewpoint.

Reviewer: There is a typo in one of the links (think this pre-dates these changes)
I believe "SClusterQL audit log export" should read "Cluster SQL audit log export"

Author: @mwang1026 The checklist follows the same order as the [Self-Hosted checklist](https://www.cockroachlabs.com/docs/v23.1/recommended-production-settings). I'm open to changing the structure tho.

For guidance on sizing, validating, and using connection pools with CockroachDB, refer [Use Connection Pools](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/connection-pooling).

## Keeping connections current

After an application establishes a connection to CockroachDB {{ site.data.products.cloud }}, those connections can occasionally become invalid. This could be due to changes in the cluster topography, rolling [upgrades]({% link cockroachcloud/upgrade-policy.md %}) and restarts, network disruptions, or cloud infrastructure unavailability.

Set the maximum lifetime of a connection to between 5 and 30 minutes. {{ site.data.products.dedicated }} and {{ site.data.products.serverless }} support 30 minutes as the maximum connection lifetime. When a node is shut down or restarted, client connections can be reset after 30 minutes, causing a disruption to applications.

> Ensure that your application can handle disruptions by implementing connection validation and [retry logic](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/transaction-retry-error-reference) appropriately. Cockroach Labs recommends using connection pools to manage the lifecycle of client connections to application servers. For guidance on connection pool sizing, connection validation, and connection retry logic, refer to [Use Connection Pools](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/connection-pooling).

Reviewer 3: we no longer recommend "connection retry logic" as of https://github.com/cockroachdb/docs/pull/17956

I think this paragraph is mixing up two different things:
- transaction retry logic is still needed. It's used when CRDB returns a serialization error (an error with a code of 40001). This is described here: https://www.cockroachlabs.com/docs/stable/transaction-retry-error-reference
- connection retry logic is something we used to recommend so an app could handle a CRDB node shutting down while the app was still connected to it. We no longer recommend that, and instead recommend using appropriate connection pool settings now.

Author: @rafiss Can I delete the paragraph since we talk about using a pool of connections and transaction retries in other sections?

So the content for this section would be:

## Keeping connections current

After an application establishes a connection to CockroachDB {{ site.data.products.cloud }}, those connections can occasionally become invalid. This could be due to changes in the cluster topography, rolling [upgrades]({% link cockroachcloud/upgrade-policy.md %}) and restarts, network disruptions, or cloud infrastructure unavailability.

Set the maximum lifetime of a connection to between 5 and 30 minutes. {{ site.data.products.dedicated }} and {{ site.data.products.serverless }} support 30 minutes as the maximum connection lifetime. When a node is shut down or restarted, client connections can be reset after 30 minutes, causing a disruption to applications.

Reviewer 3: that new paragraph looks good to me!

Author: @rafiss Updated the content. Can you verify the change and approve the PR?

docs_area: deploy
---

Before deploying CockroachDB {{ site.data.products.cloud }} in production, it is important to understand the Shared Responsibility Model that delineates the responsibilities of Cockroach Labs and the customer in managing CockroachDB {{ site.data.products.cloud }} clusters. 

Under the Shared Responsibility Model, Cockroach Labs is responsible for the following tasks: 

> - Cluster and cloud service availability and reliability

Reviewer 4: Some of the items in this and the following list are long enough that I think each item should end in a period.

- Maintenance and security of hardware and operating systems
- Database and security patches 

> - Cluster backups 

Reviewer 4: Is it worth differentiating between CRL-managed and. customer-managed backups?

Author: We do say `- (Optional) Taking customer-owned backups` in the list of customer tasks

Reviewer 4: Yes, but I guess reading it on behalf of a novice customer, I might think that seeing Cluster backups listed under CRL responsibilities means I can't take my own backup. Then I need to hold that in my mind for a bit until I make it down to the second list. It just seemed like it might be nice to add the word "Automated" or something here.

Author: Gotcha. Done!

- Scaling clusters based on workload
- Ensuring sufficient disk, compute, and memory capacity for each cluster
- Monitoring cluster health and application performance
- Ensuring that the workload is distributed appropriately across the nodes of the cluster
- Performance tuning of SQL queries and schema
- Initiating major version upgrades and selecting maintenance windows for patch releases 
- (Optional) Taking customer-owned backups

This page provides important recommendations for CockroachDB {{ site.data.products.cloud }} production tasks for which  the customer is responsible.

## Deployment options

When planning your deployment, it is important to carefully review and choose the [deployment options](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/choose-a-deployment-option) that best meet your scale, cost, security, and resiliency requirements. 

Make sure your cluster has sufficient storage, CPU, and memory to handle the workload. The general formula to calculate the storage requirement is as follows:
 
`raw data (storage, in GB) * replication factor (3 by default) * remove 40% to account for compression (0.6) * headroom (1.5-2)`

> For an example, see [Plan your Dedicated cluster]({% link cockroachcloud/plan-your-cluster.md %}).

Reviewer 4: ```suggestion
For an example, refer to [Plan your Dedicated cluster]({% link cockroachcloud/plan-your-cluster.md %}).
```
Worth searching for more instances

## Topology patterns

When planning your deployment, it is important to carefully review and choose the [topology patterns](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/topology-patterns) that best meet your latency and resiliency requirements. This is especially crucial for multi-region deployments.

## Cluster management

> You can create and manage CockroachDB {{ site.data.products.cloud }} clusters using the [Cloud Console](http://cockroachlabs.cloud), [Cloud API]({% link cockroachcloud/cloud-api.md %}), [ccloud CLI]({% link cockroachcloud/ccloud-get-started.md %}), or the [Terraform provider]({% link cockroachcloud/provision-a-cluster-with-terraform.md %}).

Reviewer 4: Is it worth making a recommendation to choose a configuration-as-code approach, since we are in the context of making recommendations?

Author: The intent here was to list the options available to the customer so they won't have to figure them out on their own. Not sure if we want to prescribe which option to use 🤔

## Network authorization

> CockroachDB {{ site.data.products.cloud }} requires you to authorize the networks that can access the cluster in order to prevent denial-of-service and brute force password attacks. During the application development phase, you might have authorized only your local machine’s network. To move into production, you need to authorize your application servers' networks.

Reviewer 4: ```suggestion
CockroachDB {{ site.data.products.cloud }} requires you to authorize the networks that can access the cluster in order to prevent denial-of-service and brute force password attacks. During the application development phase, you might have authorized only your local machine’s network. To move into production, you need to authorize your the networks used by your application servers.
```

> To verify that you have authorized the application server's network, navigate to the [Networking page]({% link cockroachcloud/connect-to-your-cluster.md %}#authorize-your-network) on the  CockroachDB {{ site.data.products.cloud }} Console and check if the application server network is listed under authorized networks. If the application server network is not on the list, authorize the network. 

Reviewer 4: ```suggestion
To verify that you have authorized an application server's network, navigate to the [Networking page]({% link cockroachcloud/connect-to-your-cluster.md %}#authorize-your-network) on the  CockroachDB {{ site.data.products.cloud }} Console and verify that the application server network is listed under **Authorized Networks**. If the network is not listed, you can add it to authorize the network. 
```

{{site.data.alerts.callout_danger}}

> While developing and testing your application on CockroachDB {{ site.data.products.dedicated }}, you may have added `0.0.0.0/0` to the allowlist, which allows all networks. CockroachDB {{ site.data.products.serverless }} allowlists `0.0.0.0/0` by default. Before moving into production, make sure you delete the `0.0.0.0/0` network.

Reviewer 4: ```suggestion
Production clusters should not authorize `0.0.0.0/0`, which allows all networks. While developing and testing your application on CockroachDB {{ site.data.products.dedicated }}, you may have manually added `0.0.0.0/0` to the allowlist. CockroachDB {{ site.data.products.serverless }} allowlists `0.0.0.0/0` by default. Before moving into production, make sure you delete the allowlist entry for the `0.0.0.0/0` network.
```

{{site.data.alerts.end}} 

> For enhanced network security and reduced network latency, you can set up private connectivity so that inbound connections to your cluster from your cloud tenant are made over the cloud provider's private network rather than over the public internet. For CockroachDB Dedicated clusters deployed on GCP, refer to [Google Cloud Platform (GCP) Virtual Private Cloud (VPC) peering]({% link cockroachcloud/network-authorization.md %}#vpc-peering). For CockroachDB Dedicated clusters or multi-region CockroachDB Serverless clusters deployed on AWS, refer to [Amazon Web Service (AWS) PrivateLink]({% link cockroachcloud/network-authorization.md %}#aws-privatelink).

Reviewer 4: ```suggestion
For enhanced network security and reduced network latency, you can set up private connectivity so that inbound connections to your cluster from your cloud tenant are made over the cloud provider's private network rather than over the public internet. For CockroachDB {{ site.data.products.dedicated }} clusters deployed on GCP, refer to [Google Cloud Platform (GCP) Virtual Private Cloud (VPC) peering]({% link cockroachcloud/network-authorization.md %}#vpc-peering). For CockroachDB {{ site.data.products.dedicated }} clusters or multi-region CockroachDB {{ site.data.products.serverless }} clusters deployed on AWS, refer to [Amazon Web Service (AWS) PrivateLink]({% link cockroachcloud/network-authorization.md %}#aws-privatelink).
```

After an application establishes a connection to CockroachDB {{ site.data.products.cloud }}, those connections can occasionally become invalid. This could be due to changes in the cluster topography, rolling [upgrades]({% link cockroachcloud/upgrade-policy.md %}) and restarts, network disruptions, or cloud infrastructure unavailability.

> Set the maximum lifetime of a connection to 5 minutes, and make sure connection validation and [retry logic](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/transaction-retry-error-reference) is used by your application. Validating and retrying connections is typically handled by the driver, framework, or the connection pool used by an application. For guidance on connection pool sizing, connection validation, and connection retry logic, see [Use Connection Pools](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/connection-pooling).

Reviewer 4: ```suggestion
Set the maximum lifetime of a connection to between 5 and 30 minutes. {{ site.data.products.dedicated }} and {{ site.data.products.serverless }} support 30 minutes as the maximum connection lifetime. When a node is shut down or restarted, client connections can be reset after 30 minutes, causing a disruption to applications.

Ensure that your application can handle disruptions by implementing connection validation and [retry logic](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/transaction-retry-error-reference) appropriately. Cockroach Labs recommends using connection pools to manage the lifecycle of client connections to application servers. For guidance on connection pool sizing, connection validation, and connection retry logic, see [Use Connection Pools](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/connection-pooling).
```
This guidance was updated in https://github.com/cockroachdb/docs/pull/17956 just yesterday. Maybe @rafiss can verify whether this suggestion is correct and helpful.

## Monitoring and alerting

> Despite CockroachDB's various [built-in safeguards](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/frequently-asked-questions#how-does-cockroachdb-survive-failures) against failure, it is critical to actively monitor the overall health and performance of a cluster running in production and to create alerting rules that promptly send notifications when there are events that require investigation or intervention.

Reviewer 4: ```suggestion
Even with CockroachDB's various [built-in safeguards](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/frequently-asked-questions#how-does-cockroachdb-survive-failures) against failure, it is critical to actively monitor the overall health and performance of a cluster running in production and to create alerting rules that promptly send notifications when there are events that require investigation or intervention.
```

To use the CockroachDB {{ site.data.products.cloud }} Console to monitor and set alerts on important events and metrics, see [Monitoring and Alerting]({% link cockroachcloud/cluster-overview-page.md %}). You can also set up monitoring with [Datadog]({% link cockroachcloud/tools-page.md %}#monitor-cockroachdb-dedicated-with-datadog) or [CloudWatch]({% link cockroachcloud/export-metrics.md %}).

## Backup and restore

> For CockroachDB {{ site.data.products.serverless }} clusters, Cockroach Labs runs [full cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#full-backups) hourly. The full backups are retained for 30 days. Once a cluster is deleted, Cockroach Labs retains the full backups for 30 days.

Reviewer 4: ```suggestion
For CockroachDB {{ site.data.products.serverless }} clusters, Cockroach Labs takes [full cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#full-backups) hourly, and retains them for 30 days. Full backups for a deleted cluster are retained for 30 days after it is deleted.
```
Does this mean we retain the last successful full backup for a deleted cluster, or we retain the 30 most recent full backups for a deleted cluster? Can a customer request that a given full backup or all full backups be manually deleted?

> For CockroachDB {{ site.data.products.dedicated }} clusters, Cockroach Labs runs [full cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#full-backups) daily and [incremental cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#incremental-backups) hourly. The full backups are retained for 30 days, while incremental backups are retained for 7 days. Once a cluster is deleted, Cockroach Labs retains the full backups for 30 days and incremental backups for 7 days. Backups are stored in the same region that a single-region cluster is running in or the primary region of a multi-region cluster.

Reviewer 4: ```suggestion
For CockroachDB {{ site.data.products.dedicated }} clusters, Cockroach Labs takes [full cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#full-backups) daily and [incremental cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#incremental-backups) hourly. Full backups are retained for 30 days, and incremental backups are retained for 7 days. Full backups for a deleted cluster are retained for 30 days, and incremental backups for 7 days.

Backups are stored in a single-region cluster's region or a multi-region cluster's primary region.
```
Does the last sentence apply equally to Dedicated and Serverless?

> You can restore cluster data to the current cluster or another cluster in the same organization. You can also restore a database or table from the Backups tab.

Reviewer 4: ```suggestion
Cluster data can be restored to the current cluster or a different cluster in the same organization. A table or database can be selectively restored from the **Backups** tab.
```

{{site.data.alerts.callout_danger}}
Restoring to a cluster will completely erase all data in the destination cluster. All cluster data will be replaced with the data from the backup. The destination cluster will be unavailable while this operation is in progress. This operation cannot be canceled, paused, or reversed.
{{site.data.alerts.end}} 

> You can [manage your own backups]({% link cockroachcloud/take-and-restore-customer-owned-backups.md %}), including incremental, database, and table level backups. To perform manual backups, you must configure either a userfile location or a cloud storage location.

Reviewer 4: ```suggestion
You can [manage your own backups]({% link cockroachcloud/take-and-restore-customer-owned-backups.md %}), including incremental, database, and table-level backups. When you perform a manual backup, you must specify a storage location, which can be on your local system or in cloud storage.
```

## Patches and upgrades

CockroachDB {{ site.data.products.cloud }} supports the latest major version of CockroachDB and the version immediately preceding it. Support for these versions includes patch version updates and security patches.

### Major version upgrades

> Major version upgrades are automatic for CockroachDB {{ site.data.products.serverless }} clusters and opt-in for CockroachDB {{ site.data.products.dedicated }} clusters. [Cluster Operators]({% link cockroachcloud/authorization.md %}#cluster-operator) must initiate major version upgrades for CockroachDB {{ site.data.products.dedicated }} clusters. When a major version upgrade is initiated for a cluster, it will upgrade to the latest patch version as well. 

Reviewer 4: ```suggestion
Major version upgrades are automatic for CockroachDB {{ site.data.products.serverless }} clusters and opt-in for CockroachDB {{ site.data.products.dedicated }} clusters. [Cluster Operators]({% link cockroachcloud/authorization.md %}#cluster-operator) must initiate major version upgrades for CockroachDB {{ site.data.products.dedicated }} clusters. When a major version upgrade is initiated for a cluster, it subsequently will be upgrade to the latest patch version automatically. 
```

> Since upgrading a cluster can have a significant impact on your workload, make sure you review the release notes for the latest version for backward compatibility, cluster setting changes, deprecations, and known limitations. Cockroach Labs recommends initiating the upgrade during off-peak periods and monitor cluster and application health post-upgrade. If you notice functional or performance regression, you will have 72 hours to roll back the changes before the upgrade is automatically finalized. Note that some new features might be unavailable until the upgrade is finalized. For more information, see [Major version upgrades]({% link cockroachcloud/upgrade-policy.md %}#major-version-upgrades).

Reviewer 4: ```suggestion
Since upgrading a cluster can have a significant impact on your workload, make sure you review the release notes for the latest version for backward compatibility, cluster setting changes, deprecations, and known limitations. Cockroach Labs recommends initiating the upgrade during off-peak periods. After the upgrade, carefully monitor cluster and application health. If you notice functional or performance regression, you can roll back the changes for up to 72 hours before the upgrade is automatically finalized. After an upgrade, some features might be unavailable until the upgrade is finalized. For more information, refer [Major version upgrades]({% link cockroachcloud/upgrade-policy.md %}#major-version-upgrades).
```

### Patch upgrades

> For CockroachDB {{ site.data.products.dedicated }} clusters, [Organization Admins]({% link cockroachcloud/authorization.md %}#org-administrator-legacy) can [set a weekly 6-hour maintenance window]({% link cockroachcloud/cluster-management.md %}#set-a-maintenance-window) during which available maintenance and patch upgrades will be applied. If no maintenance window is configured, CockroachDB {{ site.data.products.dedicated }} clusters will be automatically upgraded to the latest supported patch version as soon as it becomes available. Patch upgrades can also be [deferred for 60 days]({% link cockroachcloud/cluster-management.md %}#set-a-maintenance-window).

Reviewer 4: ```suggestion
For CockroachDB {{ site.data.products.dedicated }} clusters, [Organization Admins]({% link cockroachcloud/authorization.md %}#org-administrator-legacy) can [set a weekly 6-hour maintenance window]({% link cockroachcloud/cluster-management.md %}#set-a-maintenance-window) during which available maintenance and patch upgrades will be applied. If no maintenance window is configured, CockroachDB {{ site.data.products.dedicated }} clusters will be automatically upgraded to the latest supported patch version as soon as it becomes available. Patch upgrades can also be [deferred for 60 days]({% link cockroachcloud/cluster-management.md %}#set-a-maintenance-window).
```
I think the deferring for 60 days requires that a maintenance window is already set, right?

Author: Ah, good catch. Reordered the sentences to clarify that.

- [Network security]({% link cockroachcloud/network-authorization.md %})

You can check the status of these features on the [**PCI ready**](cluster-overview-page.html?filters=dedicated#pci-ready-dedicated-advanced) page of the CockroachDB {{ site.data.products.cloud }} Console.

## Monitoring and alerting

Despite CockroachDB's various [built-in safeguards](https://www.cockroachlabs.com/docs/{{site.current_cloud_version}}/frequently-asked-questions#how-does-cockroachdb-survive-failures) against failure, it is critical to actively monitor the overall health and performance of a cluster running in production and to create alerting rules that promptly send notifications when there are events that require investigation or intervention.

To use the CockroachDB {{ site.data.products.cloud }} Console to monitor and set alerts on important events and metrics, see [Monitoring and Alerting]({% link cockroachcloud/cluster-overview-page.md %}). You can also set up monitoring with [Datadog]({% link cockroachcloud/tools-page.md %}#monitor-cockroachdb-dedicated-with-datadog) or [CloudWatch]({% link cockroachcloud/export-metrics.md %}).

## Backup and restore

> For CockroachDB {{ site.data.products.serverless }} clusters, Cockroach Labs runs [full cluster backups]({% link cockroachcloud/use-managed-service-backups.md %}/#full-backups) hourly. The full backups are retained for 30 days. Once a cluster is deleted, Cockroach Labs retains the full backups for 30 days.

Reviewer 5: ```suggestion
For CockroachDB {{ site.data.products.serverless }} clusters, Cockroach Labs runs [full cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#full-backups) hourly. The full backups are retained for 30 days. Once a cluster is deleted, Cockroach Labs retains the full backups for 30 days.
```

> For CockroachDB {{ site.data.products.dedicated }} clusters, Cockroach Labs runs [full cluster backups]({% link cockroachcloud/use-managed-service-backups.md %}/#full-backups) daily and [incremental cluster backups]({% link cockroachcloud/use-managed-service-backups.md %}/incremental-backups) hourly. The full backups are retained for 30 days, while incremental backups are retained for 7 days. Once a cluster is deleted, Cockroach Labs retains the full backups for 30 days and incremental backups for 7 days. Backups are stored in the same region that a single-region cluster is running in or the primary region of a multi-region cluster.

Reviewer 5: ```suggestion
For CockroachDB {{ site.data.products.dedicated }} clusters, Cockroach Labs runs [full cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#full-backups) daily and [incremental cluster backups](https://www.cockroachlabs.com/docs/{{ site.current_cloud_version }}/take-full-and-incremental-backups#incremental-backups) hourly. The full backups are retained for 30 days, while incremental backups are retained for 7 days. Once a cluster is deleted, Cockroach Labs retains the full backups for 30 days and incremental backups for 7 days. Backups are stored in the same region that a single-region cluster is running in or the primary region of a multi-region cluster.
```

{{site.data.alerts.end}} 

You can [manage your own backups]({% link cockroachcloud/take-and-restore-customer-owned-backups.md %}), including incremental, database, and table level backups. To perform manual backups, you must configure either a userfile location or a cloud storage location.

## Patches and upgrades

CockroachDB {{ site.data.products.cloud }} supports the latest major version of CockroachDB and the version immediately preceding it. Support for these versions includes patch version updates and security patches.

### Major version upgrades

Major version upgrades are automatic for CockroachDB {{ site.data.products.serverless }} clusters and opt-in for CockroachDB {{ site.data.products.dedicated }} clusters. [Cluster Operators]({% link cockroachcloud/authorization.md %}#cluster-operator) must initiate major version upgrades for CockroachDB {{ site.data.products.dedicated }} clusters. When a major version upgrade is initiated for a cluster, it will upgrade to the latest patch version as well. 

> Since upgrading a cluster can have a significant impact on your workload, make sure you review the release notes for the latest version for backward compatibility, cluster setting changes, deprecations, and known limitations. Cockroach Labs recommends initiating the upgrade during off-peak periods and monitor cluster and application health post-upgrade. If you notice functional or performance regression, you can roll back the upgrade approximately 72 hours before the upgrade is automatically finalized. Note that some new features might be unavailable until the upgrade is finalized. For more information, see [Major version upgrades]({% link cockroachcloud/upgrade-policy.md %}#major-version-upgrades).

Reviewer: "you can roll back the upgrade approximately 72 hours before the upgrade is automatically finalized" should be something like  "you will have 72 hours to roll back the changes before the upgrade is automatically finalized"
